Index: linux-tx8m-1620/drivers/staging/media/imx/dwc-mipi-csi2.c
===================================================================
--- linux-tx8m-1620.orig/drivers/staging/media/imx/dwc-mipi-csi2.c
+++ linux-tx8m-1620/drivers/staging/media/imx/dwc-mipi-csi2.c
@@ -396,6 +396,15 @@ static const struct csi2h_pix_format dwc
 		.code = MEDIA_BUS_FMT_SBGGR8_1X8,
 		.fmt_reg = 0x2A,
 	}, {
+		.code = MEDIA_BUS_FMT_SRGGB8_1X8,
+		.fmt_reg = 0x2A,
+	}, {
+		.code = MEDIA_BUS_FMT_SGBRG8_1X8,
+		.fmt_reg = 0x2A,
+	}, {
+		.code = MEDIA_BUS_FMT_SGRBG8_1X8,
+		.fmt_reg = 0x2A,
+	}, {
 		.code = MEDIA_BUS_FMT_SBGGR10_1X10,
 		.fmt_reg = 0x2B,
 	}, {
@@ -702,6 +711,9 @@ static void disp_mix_gasket_config(struc
 		fmt_val = DT_YUV422_8;
 		break;
 	case MEDIA_BUS_FMT_SBGGR8_1X8:
+	case MEDIA_BUS_FMT_SRGGB8_1X8:
+	case MEDIA_BUS_FMT_SGBRG8_1X8:
+	case MEDIA_BUS_FMT_SGRBG8_1X8:
 		fmt_val = DT_RAW8;
 		break;
 	case MEDIA_BUS_FMT_SBGGR10_1X10:
