Index: linux-qs93/drivers/net/ethernet/stmicro/stmmac/stmmac_mdio.c
===================================================================
--- linux-qs93.orig/drivers/net/ethernet/stmicro/stmmac/stmmac_mdio.c
+++ linux-qs93/drivers/net/ethernet/stmicro/stmmac/stmmac_mdio.c
@@ -358,7 +358,7 @@ int stmmac_mdio_reset(struct mii_bus *bu
 
 		reset_gpio = devm_gpiod_get_optional(priv->device,
 						     "snps,reset",
-						     GPIOD_OUT_LOW);
+						     GPIOD_OUT_HIGH);
 		if (IS_ERR_OR_NULL(reset_gpio)) {
 			ret = PTR_ERR_OR_ZERO(reset_gpio);
 			goto mdio_gpio_reset_end;
Index: linux-qs93/drivers/net/phy/mdio_bus.c
===================================================================
--- linux-qs93.orig/drivers/net/phy/mdio_bus.c
+++ linux-qs93/drivers/net/phy/mdio_bus.c
@@ -9,6 +9,7 @@
 #define pr_fmt(fmt) KBUILD_MODNAME ": " fmt
 
 #include <linux/acpi.h>
+#include <linux/clk.h>
 #include <linux/delay.h>
 #include <linux/device.h>
 #include <linux/errno.h>
@@ -531,6 +532,7 @@ int __mdiobus_register(struct mii_bus *b
 	struct mdio_device *mdiodev;
 	int i, err;
 	struct gpio_desc *gpiod;
+	struct clk *clk;
 
 	if (NULL == bus || NULL == bus->name ||
 	    NULL == bus->read || NULL == bus->write)
@@ -566,6 +568,10 @@ int __mdiobus_register(struct mii_bus *b
 	mutex_init(&bus->mdio_lock);
 	mutex_init(&bus->shared_lock);
 
+	clk = devm_clk_get_optional_enabled(&bus->dev, NULL);
+	if (IS_ERR(clk))
+		return PTR_ERR(clk);
+
 	/* assert bus level PHY GPIO reset */
 	gpiod = devm_gpiod_get_optional(&bus->dev, "reset", GPIOD_OUT_HIGH);
 	if (IS_ERR(gpiod)) {
