diff --git a/board/karo/tx8m/tx8mp.c b/board/karo/tx8m/tx8mp.c
index 4a7aebc86a..4f04ce808c 100644
--- a/board/karo/tx8m/tx8mp.c
+++ b/board/karo/tx8m/tx8mp.c
@@ -491,12 +491,17 @@ int ft_board_setup(void *blob, bd_t *bd)
 		fdt_addr_t addr;
 		bool ocpol, pwrpol;
 
+		if (node <= 0) {
+			debug("%s alias not found\n", usb_aliases[i]);
+			continue;
+		}
 		if (!fdtdec_get_is_enabled(blob, node)) {
 			printf("%s: '%s' device is not enabled\n", __func__,
 			       usb_aliases[i]);
 			continue;
 		}
-		addr = fdtdec_get_addr(blob, node, "reg");
+		addr = fdtdec_get_addr_size_auto_noparent(blob, node,
+							  "reg", 0, NULL, true);
 		if (IS_ERR_VALUE(addr)) {
 			printf("Failed to get address of '%s' device: %d\n",
 			       usb_aliases[i], ret);
