diff --git a/arch/arm/dts/Makefile b/arch/arm/dts/Makefile
index 8084adec770..7ce6991230e 100644
--- a/arch/arm/dts/Makefile
+++ b/arch/arm/dts/Makefile
@@ -792,8 +792,10 @@ dtb-$(CONFIG_ARCH_IMX8M) += \
 	imx8mm-verdin.dtb \
 	imx8mm-qs8m-mq00.dtb \
 	imx8mm-qs8m-mq00-qsbase2.dtb \
+	imx8mm-qs8m-mq00-qsbase4.dtb \
 	imx8mm-qsxm-mm60.dtb \
 	imx8mm-qsxm-mm60-qsbase3.dtb \
+	imx8mm-qsxm-mm60-qsbase4.dtb \
 	imx8mm-tx8m-1610-mb7.dtb \
 	imx8mm-tx8m-1610.dtb \
 	imx8mm-tx8m-1620.dtb \
@@ -803,9 +805,11 @@ dtb-$(CONFIG_ARCH_IMX8M) += \
 	imx8mn-evk.dtb \
 	imx8mn-qs8m-nd00.dtb \
 	imx8mn-qs8m-nd00-qsbase2.dtb \
+	imx8mn-qs8m-nd00-qsbase4.dtb \
 	imx8mn-tx8m-nd00.dtb \
 	imx8mp-qsxp-ml81.dtb \
 	imx8mp-qsxp-ml81-qsbase3.dtb \
+	imx8mp-qsxp-ml81-qsbase4.dtb \
 	imx8mp-tx8p-ml81.dtb \
 	imx8mp-tx8p-ml81-mipi-mb.dtb \
 	imx8mq-ddr3l-val.dtb \
