Index: imx-atf-qsxp/plat/imx/imx8m/imx8mp/include/platform_def.h
===================================================================
--- imx-atf-qsxp.orig/plat/imx/imx8m/imx8mp/include/platform_def.h
+++ imx-atf-qsxp/plat/imx/imx8m/imx8mp/include/platform_def.h
@@ -40,7 +40,9 @@
 #define BL2_BASE			U(0x970000)
 #define BL2_SIZE			SZ_128K
 #define BL2_LIMIT			(BL2_BASE + BL2_SIZE)
+#ifndef BL31_BASE
 #define BL31_BASE			U(0x950000)
+#endif
 #define IMX_FIP_BASE			U(0x40310000)
 #define IMX_FIP_SIZE			U(0x000300000)
 #define IMX_FIP_LIMIT			U(FIP_BASE + FIP_SIZE)
@@ -50,10 +52,14 @@
 
 #define PLAT_IMX8MP_BOOT_MMC_BASE	U(0x30B50000) /* SD */
 #else
+#ifndef BL31_BASE
 #define BL31_BASE			U(0x970000)
 #endif
+#endif
 
+#ifndef BL31_SIZE
 #define BL31_SIZE			SZ_128K
+#endif
 #define BL31_LIMIT			(BL31_BASE + BL31_SIZE)
 
 #define PLAT_PRI_BITS			U(3)
@@ -170,7 +176,7 @@
 #define SRC_GPR1_OFFSET			U(0x74)
 
 #define SRC_SCR_M4_ENABLE_MASK		BIT(3)
-#define SRC_SCR_M4C_NON_SCLR_RST_MASK  	BIT(0)
+#define SRC_SCR_M4C_NON_SCLR_RST_MASK	BIT(0)
 
 #define SNVS_LPCR			U(0x38)
 #define SNVS_LPCR_SRTC_ENV		BIT(0)
@@ -196,7 +202,9 @@
 #define OCRAM_S_BASE			U(0x00180000)
 #define OCRAM_S_SIZE			U(0x8000)
 #define OCRAM_S_LIMIT			(OCRAM_S_BASE + OCRAM_S_SIZE)
+#ifndef SAVED_DRAM_TIMING_BASE
 #define SAVED_DRAM_TIMING_BASE		OCRAM_S_BASE
+#endif
 
 #define COUNTER_FREQUENCY		8000000 /* 8MHz */
 
Index: imx-atf-qsxp/plat/imx/imx8m/imx8mp/platform.mk
===================================================================
--- imx-atf-qsxp.orig/plat/imx/imx8m/imx8mp/platform.mk
+++ imx-atf-qsxp/plat/imx/imx8m/imx8mp/platform.mk
@@ -150,12 +150,21 @@ ERRATA_A53_835769	:=	1
 ERRATA_A53_843419	:=	1
 ERRATA_A53_855873	:=	1
 
+BL31_BASE		?=	0x970000
+$(eval $(call add_define,BL31_BASE))
+
+BL31_SIZE		?=	0x20000
+$(eval $(call add_define,BL31_SIZE))
+
 BL32_BASE		?=	0x56000000
 $(eval $(call add_define,BL32_BASE))
 
 BL32_SIZE		?=	0x2000000
 $(eval $(call add_define,BL32_SIZE))
 
+SAVED_DRAM_TIMING_BASE ?= 0x180000
+$(eval $(call add_define,SAVED_DRAM_TIMING_BASE))
+
 IMX_BOOT_UART_BASE	?=	0x30890000
 ifeq (${IMX_BOOT_UART_BASE},auto)
     override IMX_BOOT_UART_BASE	:=	0
