diff --git a/iMX8M/soc.mak b/iMX8M/soc.mak
index 4bfdc37..762bf89 100644
--- a/iMX8M/soc.mak
+++ b/iMX8M/soc.mak
@@ -143,7 +143,7 @@ dtb = evk.dtb
 $(dtb):
 	./$(DTB_PREPROC) $(PLAT)-evk.dtb $(dtb) $(dtbs)
 
-u-boot.itb: $(dtb)
+u-boot.itb: $(dtb) u-boot-nodtb.bin tee.bin bl31.bin
 	./$(PAD_IMAGE) tee.bin
 	./$(PAD_IMAGE) bl31.bin
 	./$(PAD_IMAGE) u-boot-nodtb.bin $(dtb)
@@ -155,7 +155,7 @@ dtb_ddr3l = valddr3l.dtb
 $(dtb_ddr3l):
 	./$(DTB_PREPROC) $(PLAT)-ddr3l-$(VAL_BOARD).dtb $(dtb_ddr3l) $(dtbs)
 
-u-boot-ddr3l.itb: $(dtb_ddr3l)
+u-boot-ddr3l.itb: $(dtb_ddr3l) u-boot-nodtb.bin tee.bin bl31.bin
 	./$(PAD_IMAGE) tee.bin
 	./$(PAD_IMAGE) bl31.bin
 	./$(PAD_IMAGE) u-boot-nodtb.bin $(dtb_ddr3l)
@@ -179,7 +179,7 @@ dtb_ddr4 = valddr4.dtb
 $(dtb_ddr4):
 	./$(DTB_PREPROC) $(PLAT)-ddr4-$(VAL_BOARD).dtb $(dtb_ddr4) $(dtbs)
 
-u-boot-ddr4.itb: $(dtb_ddr4)
+u-boot-ddr4.itb: $(dtb_ddr4) u-boot-nodtb.bin tee.bin bl31.bin
 	./$(PAD_IMAGE) tee.bin
 	./$(PAD_IMAGE) bl31.bin
 	./$(PAD_IMAGE) u-boot-nodtb.bin $(dtb_ddr4)
@@ -191,7 +191,7 @@ dtb_ddr4_evk = evkddr4.dtb
 $(dtb_ddr4_evk):
 	./$(DTB_PREPROC) $(PLAT)-ddr4-evk.dtb $(dtb_ddr4_evk) $(dtbs)
 
-u-boot-ddr4-evk.itb: $(dtb_ddr4_evk)
+u-boot-ddr4-evk.itb: $(dtb_ddr4_evk) u-boot-nodtb.bin tee.bin bl31.bin
 	./$(PAD_IMAGE) tee.bin
 	./$(PAD_IMAGE) bl31.bin
 	./$(PAD_IMAGE) u-boot-nodtb.bin $(dtb_ddr4_evk)
