Index: linux-imx/drivers/mfd/Kconfig
===================================================================
--- linux-imx.orig/drivers/mfd/Kconfig
+++ linux-imx/drivers/mfd/Kconfig
@@ -1993,6 +1993,8 @@ config MFD_PCA9450
 	bool "PCA9450 Power Management chip"
 	depends on I2C=y
 	select MFD_CORE
+	select REGMAP_I2C
+	select REGMAP_IRQ
 	help
 	  if you say yes here you get support for the PCA9450
 	  Power Management chips.
Index: linux-imx/drivers/regulator/Kconfig
===================================================================
--- linux-imx.orig/drivers/regulator/Kconfig
+++ linux-imx/drivers/regulator/Kconfig
@@ -679,10 +679,10 @@ config REGULATOR_PCAP
 	 PCAP2 PMIC.
 
 config REGULATOR_PCA9450
-    tristate "NXP PCA9450 Power Regulator"
-    depends on MFD_PCA9450
-    help
-      This driver supports PCA9450 voltage regulator chips.
+	tristate "NXP PCA9450 Power Regulator"
+	depends on MFD_PCA9450
+	help
+	  This driver supports PCA9450 voltage regulator chips.
 
 config REGULATOR_PCF50633
 	tristate "NXP PCF50633 regulator driver"
