diff --git a/cmd/fdt.c b/cmd/fdt.c
index 2a207bf2b5..2c6dccd895 100644
--- a/cmd/fdt.c
+++ b/cmd/fdt.c
@@ -31,6 +31,11 @@ static int fdt_parse_prop(char *const*newval, int count, char *data, int *len);
 static int fdt_print(const char *pathp, char *prop, int depth);
 static int is_printable_string(const void *data, int len);
 
+int _fdt_print(void)
+{
+	fdt_print("/", NULL, 32);
+}
+
 /*
  * The working_fdt points to our working flattened device tree.
  */
diff --git a/include/fdt.h b/include/fdt.h
index b97b3f2bcc..4f20cf4901 100644
--- a/include/fdt.h
+++ b/include/fdt.h
@@ -1 +1,3 @@
 #include "../scripts/dtc/libfdt/fdt.h"
+
+int _fdt_print(void);
