Index: imx-atf-karo/plat/imx/imx91/include/platform_def.h
===================================================================
--- imx-atf-karo.orig/plat/imx/imx91/include/platform_def.h
+++ imx-atf-karo/plat/imx/imx91/include/platform_def.h
@@ -30,11 +30,8 @@
 #define PLAT_MAX_OFF_STATE		U(4)
 #define PLAT_MAX_RET_STATE		U(2)
 
-#define BL31_BASE			U(0x204C0000)
 #define BL31_LIMIT			U(0x204E0000)
 
-#define SAVED_DRAM_TIMING_BASE		U(0x2049C000)
-
 /* non-secure uboot base */
 /* TODO */
 #define PLAT_NS_IMAGE_OFFSET		U(0x80200000)
Index: imx-atf-karo/plat/imx/imx91/platform.mk
===================================================================
--- imx-atf-karo.orig/plat/imx/imx91/platform.mk
+++ imx-atf-karo/plat/imx/imx91/platform.mk
@@ -62,3 +62,17 @@ BL32_BASE               ?=      0x960000
 BL32_SIZE               ?=      0x02000000
 $(eval $(call add_define,BL32_BASE))
 $(eval $(call add_define,BL32_SIZE))
+
+BL31_BASE		?=	0x204C0000
+BL31_SIZE		?=	0x1D000
+$(eval $(call add_define,BL31_BASE))
+$(eval $(call add_define,BL31_SIZE))
+
+SAVED_DRAM_TIMING_BASE	?=	0x2049C000
+$(eval $(call add_define,SAVED_DRAM_TIMING_BASE))
+
+IMX_LPUART_BASE	?=	0x44380000
+ifeq (${IMX_LPUART_BASE},auto)
+    override IMX_LPUART_BASE	:=	0
+endif
+$(eval $(call add_define,IMX_LPUART_BASE))
Index: imx-atf-karo/plat/imx/imx93/include/platform_def.h
===================================================================
--- imx-atf-karo.orig/plat/imx/imx93/include/platform_def.h
+++ imx-atf-karo/plat/imx/imx93/include/platform_def.h
@@ -28,15 +28,12 @@
 #define PLAT_MAX_OFF_STATE		U(4)
 #define PLAT_MAX_RET_STATE		U(2)
 
-#define BL31_BASE			U(0x204E0000)
 #define BL31_LIMIT			U(0x20520000)
 
-#define SAVED_DRAM_TIMING_BASE		U(0x2051c000)
-
 /* non-secure uboot base */
 /* TODO */
 #define PLAT_NS_IMAGE_OFFSET		U(0x80200000)
-#define BL32_FDT_OVERLAY_ADDR           (PLAT_NS_IMAGE_OFFSET + 0x3000000)
+#define BL32_FDT_OVERLAY_ADDR		(PLAT_NS_IMAGE_OFFSET + 0x3000000)
 
 /* GICv4 base address */
 #define PLAT_GICD_BASE			U(0x48000000)
Index: imx-atf-karo/plat/imx/imx93/platform.mk
===================================================================
--- imx-atf-karo.orig/plat/imx/imx93/platform.mk
+++ imx-atf-karo/plat/imx/imx93/platform.mk
@@ -63,3 +63,11 @@ BL32_BASE               ?=      0x960000
 BL32_SIZE               ?=      0x02000000
 $(eval $(call add_define,BL32_BASE))
 $(eval $(call add_define,BL32_SIZE))
+
+BL31_BASE		?=	0x204E0000
+BL31_SIZE		?=	0x25000
+$(eval $(call add_define,BL31_BASE))
+$(eval $(call add_define,BL31_SIZE))
+
+SAVED_DRAM_TIMING_BASE	?=	0x2051C000
+$(eval $(call add_define,SAVED_DRAM_TIMING_BASE))
Index: imx-atf-karo/plat/imx/imx8m/imx8mm/include/platform_def.h
===================================================================
--- imx-atf-karo.orig/plat/imx/imx8m/imx8mm/include/platform_def.h
+++ imx-atf-karo/plat/imx/imx8m/imx8mm/include/platform_def.h
@@ -52,11 +52,8 @@
 #define IMX_FIP_MMC_BASE		U(0x100000)
 
 #define PLAT_IMX8MM_BOOT_MMC_BASE	U(0x30B50000) /* SD */
-#else
-#define BL31_BASE			U(0x920000)
 #endif
 
-#define BL31_SIZE			SZ_128K
 #define BL31_LIMIT			(BL31_BASE + BL31_SIZE)
 
 /* non-secure uboot base */
@@ -153,7 +150,7 @@
 #define SRC_GPR10_PERSIST_SECONDARY_BOOT	BIT(30)
 
 #define SRC_SCR_M4_ENABLE_MASK		BIT(3)
-#define SRC_SCR_M4C_NON_SCLR_RST_MASK  	BIT(0)
+#define SRC_SCR_M4C_NON_SCLR_RST_MASK	BIT(0)
 #define IMX_M4_STATUS			(IMX_SRC_BASE + SRC_M4RCR)
 #define IMX_M4_ENABLED_MASK		SRC_SCR_M4C_NON_SCLR_RST_MASK
 #define LPA_STATUS			U(0x94)
@@ -177,7 +174,6 @@
 #define OCRAM_S_BASE			U(0x00180000)
 #define OCRAM_S_SIZE			U(0x8000)
 #define OCRAM_S_LIMIT			(OCRAM_S_BASE + OCRAM_S_SIZE)
-#define SAVED_DRAM_TIMING_BASE		OCRAM_S_BASE
 
 #define COUNTER_FREQUENCY		8000000 /* 8MHz */
 
Index: imx-atf-karo/plat/imx/imx8m/imx8mn/include/platform_def.h
===================================================================
--- imx-atf-karo.orig/plat/imx/imx8m/imx8mn/include/platform_def.h
+++ imx-atf-karo/plat/imx/imx8m/imx8mn/include/platform_def.h
@@ -40,8 +40,6 @@
 #define PLAT_SDEI_NORMAL_PRI		0x20
 #define PLAT_SDEI_SGI_PRIVATE		U(9)
 
-#define BL31_BASE			U(0x960000)
-#define BL31_SIZE			SZ_128K
 #define BL31_LIMIT			(BL31_BASE + BL31_SIZE)
 
 /* non-secure uboot base */
@@ -131,7 +129,7 @@
 #define SRC_GPR1_OFFSET			U(0x74)
 
 #define SRC_SCR_M4_ENABLE_MASK		BIT(3)
-#define SRC_SCR_M4C_NON_SCLR_RST_MASK  	BIT(0)
+#define SRC_SCR_M4C_NON_SCLR_RST_MASK	BIT(0)
 
 #define SNVS_LPCR			U(0x38)
 #define SNVS_LPCR_SRTC_ENV		BIT(0)
@@ -158,7 +156,6 @@
 #define OCRAM_S_BASE			U(0x00180000)
 #define OCRAM_S_SIZE			U(0x8000)
 #define OCRAM_S_LIMIT			(OCRAM_S_BASE + OCRAM_S_SIZE)
-#define SAVED_DRAM_TIMING_BASE		OCRAM_S_BASE
 
 #define COUNTER_FREQUENCY		8000000 /* 8MHz */
 
Index: imx-atf-karo/plat/imx/imx8m/imx8mp/include/platform_def.h
===================================================================
--- imx-atf-karo.orig/plat/imx/imx8m/imx8mp/include/platform_def.h
+++ imx-atf-karo/plat/imx/imx8m/imx8mp/include/platform_def.h
@@ -49,11 +49,8 @@
 #define IMX_FIP_MMC_BASE		U(0x100000)
 
 #define PLAT_IMX8MP_BOOT_MMC_BASE	U(0x30B50000) /* SD */
-#else
-#define BL31_BASE			U(0x970000)
 #endif
 
-#define BL31_SIZE			SZ_128K
 #define BL31_LIMIT			(BL31_BASE + BL31_SIZE)
 
 #define PLAT_PRI_BITS			U(3)
@@ -170,7 +167,7 @@
 #define SRC_GPR1_OFFSET			U(0x74)
 
 #define SRC_SCR_M4_ENABLE_MASK		BIT(3)
-#define SRC_SCR_M4C_NON_SCLR_RST_MASK  	BIT(0)
+#define SRC_SCR_M4C_NON_SCLR_RST_MASK	BIT(0)
 
 #define SNVS_LPCR			U(0x38)
 #define SNVS_LPCR_SRTC_ENV		BIT(0)
@@ -196,7 +193,6 @@
 #define OCRAM_S_BASE			U(0x00180000)
 #define OCRAM_S_SIZE			U(0x8000)
 #define OCRAM_S_LIMIT			(OCRAM_S_BASE + OCRAM_S_SIZE)
-#define SAVED_DRAM_TIMING_BASE		OCRAM_S_BASE
 
 #define COUNTER_FREQUENCY		8000000 /* 8MHz */
 
Index: imx-atf-karo/plat/imx/imx8m/imx8mq/include/platform_def.h
===================================================================
--- imx-atf-karo.orig/plat/imx/imx8m/imx8mq/include/platform_def.h
+++ imx-atf-karo/plat/imx/imx8m/imx8mq/include/platform_def.h
@@ -122,7 +122,7 @@
 #define SRC_GPR10_PERSIST_SECONDARY_BOOT	BIT(30)
 
 #define SRC_SCR_M4_ENABLE_MASK		BIT(3)
-#define SRC_SCR_M4C_NON_SCLR_RST_MASK  	BIT(0)
+#define SRC_SCR_M4C_NON_SCLR_RST_MASK	BIT(0)
 #define IMX_M4_STATUS			(IMX_SRC_BASE + SRC_M4RCR)
 #define IMX_M4_ENABLED_MASK		SRC_SCR_M4C_NON_SCLR_RST_MASK
 #define LPA_STATUS			U(0x94)
@@ -150,7 +150,7 @@
 #define OCRAM_S_LIMIT			(OCRAM_S_BASE + OCRAM_S_SIZE)
 
 #define COUNTER_FREQUENCY		8333333 /* 25MHz / 3 */
-#define IMX_SEPARATE_NOBITS_BASE 	OCRAM_S_BASE
+#define IMX_SEPARATE_NOBITS_BASE	OCRAM_S_BASE
 #define IMX_SEPARATE_NOBITS_LIMIT	OCRAM_S_LIMIT
 #define IMX_SEPARATE_COHERENT_MEM_BASE  U(0x90E000)
 #define IMX_SEPARATE_COHERENT_MEM_LIMIT U(0x90F000)
@@ -160,8 +160,6 @@
 #ifdef IMX_ANDROID_BUILD
 #if !DEBUG_CONSOLE
 #undef LOG_LEVEL
-#define LOG_LEVEL 			0
+#define LOG_LEVEL			0
 #endif
 #endif
-
-#define IMX_WDOG_B_RESET
Index: imx-atf-karo/plat/imx/imx8m/imx8mm/platform.mk
===================================================================
--- imx-atf-karo.orig/plat/imx/imx8m/imx8mm/platform.mk
+++ imx-atf-karo/plat/imx/imx8m/imx8mm/platform.mk
@@ -161,6 +161,14 @@ $(eval $(call add_define,BL32_BASE))
 BL32_SIZE		?=	0x2000000
 $(eval $(call add_define,BL32_SIZE))
 
+BL31_BASE		?=	0x920000
+BL31_SIZE		?=	0x20000
+$(eval $(call add_define,BL31_BASE))
+$(eval $(call add_define,BL31_SIZE))
+
+SAVED_DRAM_TIMING_BASE ?= 0x180000
+$(eval $(call add_define,SAVED_DRAM_TIMING_BASE))
+
 IMX_BOOT_UART_BASE	?=	0x30890000
 ifeq (${IMX_BOOT_UART_BASE},auto)
     override IMX_BOOT_UART_BASE	:=	0
Index: imx-atf-karo/plat/imx/imx8m/imx8mn/platform.mk
===================================================================
--- imx-atf-karo.orig/plat/imx/imx8m/imx8mn/platform.mk
+++ imx-atf-karo/plat/imx/imx8m/imx8mn/platform.mk
@@ -67,6 +67,14 @@ $(eval $(call add_define,BL32_BASE))
 BL32_SIZE		?=	0x2000000
 $(eval $(call add_define,BL32_SIZE))
 
+BL31_BASE		?=	0x960000
+BL31_SIZE		?=	0x20000
+$(eval $(call add_define,BL31_BASE))
+$(eval $(call add_define,BL31_SIZE))
+
+SAVED_DRAM_TIMING_BASE ?= 0x180000
+$(eval $(call add_define,SAVED_DRAM_TIMING_BASE))
+
 IMX_BOOT_UART_BASE	?=	0x30890000
 ifeq (${IMX_BOOT_UART_BASE},auto)
     override IMX_BOOT_UART_BASE	:=	0
Index: imx-atf-karo/plat/imx/imx8m/imx8mp/platform.mk
===================================================================
--- imx-atf-karo.orig/plat/imx/imx8m/imx8mp/platform.mk
+++ imx-atf-karo/plat/imx/imx8m/imx8mp/platform.mk
@@ -158,6 +158,14 @@ $(eval $(call add_define,BL32_BASE))
 BL32_SIZE		?=	0x2000000
 $(eval $(call add_define,BL32_SIZE))
 
+BL31_BASE		?=	0x970000
+BL31_SIZE		?=	0x20000
+$(eval $(call add_define,BL31_BASE))
+$(eval $(call add_define,BL31_SIZE))
+
+SAVED_DRAM_TIMING_BASE ?= 0x180000
+$(eval $(call add_define,SAVED_DRAM_TIMING_BASE))
+
 IMX_BOOT_UART_BASE	?=	0x30890000
 ifeq (${IMX_BOOT_UART_BASE},auto)
     override IMX_BOOT_UART_BASE	:=	0
