diff --git a/arch/arm64/boot/dts/freescale/imx93.dtsi b/arch/arm64/boot/dts/freescale/imx93.dtsi
index 13ab4797e946..4cf48f6bcf41 100644
--- a/arch/arm64/boot/dts/freescale/imx93.dtsi
+++ b/arch/arm64/boot/dts/freescale/imx93.dtsi
@@ -107,6 +107,18 @@ clk_ext1: clock-ext1 {
 		clock-output-names = "clk_ext1";
 	};
 
+	mqs1: mqs1 {
+		compatible = "fsl,imx93-mqs";
+		gpr = <&anomix_ns_gpr>;
+		status = "disabled";
+	};
+
+	mqs2: mqs2 {
+		compatible = "fsl,imx93-mqs";
+		gpr = <&wakeupmix_gpr>;
+		status = "disabled";
+	};
+
 	pmu {
 		compatible = "arm,cortex-a55-pmu";
 		interrupts = <GIC_PPI 7 (GIC_CPU_MASK_SIMPLE(2) | IRQ_TYPE_LEVEL_HIGH)>;
@@ -138,6 +150,18 @@ gic: interrupt-controller@48000000 {
 		interrupt-parent = <&gic>;
 	};
 
+	usbphynop1: usbphynop1 {
+		compatible = "usb-nop-xceiv";
+		clocks = <&clk IMX93_CLK_USB_PHY_BURUNIN>;
+		clock-names = "main_clk";
+	};
+
+	usbphynop2: usbphynop2 {
+		compatible = "usb-nop-xceiv";
+		clocks = <&clk IMX93_CLK_USB_PHY_BURUNIN>;
+		clock-names = "main_clk";
+	};
+
 	thermal-zones {
 		cpu-thermal {
 			polling-delay-passive = <250>;
@@ -190,7 +214,7 @@ anomix_ns_gpr: syscon@44210000 {
 			};
 
 			mu1: mailbox@44230000 {
-				compatible = "fsl,imx93-mu", "fsl,imx8ulp-mu";
+				compatible = "fsl,imx93-mu-s4";
 				reg = <0x44230000 0x10000>;
 				interrupts = <GIC_SPI 22 IRQ_TYPE_LEVEL_HIGH>;
 				clocks = <&clk IMX93_CLK_MU1_B_GATE>;
@@ -281,6 +305,25 @@ system_counter: timer@44290000 {
 				nxp,no-divider;
 			};
 
+			tpm1: pwm@44310000 {
+				compatible = "fsl,imx7ulp-pwm";
+				reg = <0x44310000 0x1000>;
+				clocks = <&clk IMX93_CLK_TPM1_GATE>;
+				#pwm-cells = <3>;
+				status = "disabled";
+			};
+
+			tpm2: pwm@44320000 {
+				compatible = "fsl,imx7ulp-pwm";
+				reg = <0x44320000 0x1000>;
+				clocks = <&clk IMX93_CLK_TPM2_GATE>;
+				assigned-clocks = <&clk IMX93_CLK_TPM2>;
+				assigned-clock-parents = <&clk IMX93_CLK_24M>;
+				assigned-clock-rates = <24000000>;
+				#pwm-cells = <3>;
+				status = "disabled";
+			};
+
 			i3c1: i3c-master@44330000 {
 				#address-cells = <3>;
 				#size-cells = <0>;
@@ -351,7 +398,7 @@ lpspi2: spi@44370000 {
 			};
 
 			lpuart1: serial@44380000 {
-				compatible = "fsl,imx93-lpuart", "fsl,imx8ulp-lpuart", "fsl,imx7ulp-lpuart";
+				compatible = "fsl,imx93-lpuart", "fsl,imx8ulp-lpuart";
 				reg = <0x44380000 0x1000>;
 				interrupts = <GIC_SPI 19 IRQ_TYPE_LEVEL_HIGH>;
 				clocks = <&clk IMX93_CLK_LPUART1_GATE>;
@@ -360,7 +407,7 @@ lpuart1: serial@44380000 {
 			};
 
 			lpuart2: serial@44390000 {
-				compatible = "fsl,imx93-lpuart", "fsl,imx8ulp-lpuart", "fsl,imx7ulp-lpuart";
+				compatible = "fsl,imx93-lpuart", "fsl,imx8ulp-lpuart";
 				reg = <0x44390000 0x1000>;
 				interrupts = <GIC_SPI 20 IRQ_TYPE_LEVEL_HIGH>;
 				clocks = <&clk IMX93_CLK_LPUART2_GATE>;
@@ -398,12 +445,6 @@ sai1: sai@443b0000 {
 				status = "disabled";
 			};
 
-			mqs1: mqs1 {
-				compatible = "fsl,imx93-mqs";
-				gpr = <&anomix_ns_gpr>;
-				status = "disabled";
-			};
-
 			iomuxc: pinctrl@443c0000 {
 				compatible = "fsl,imx93-iomuxc";
 				reg = <0x443c0000 0x10000>;
@@ -669,7 +710,7 @@ wakeupmix_gpr: syscon@42420000 {
 			};
 
 			mu2: mailbox@42440000 {
-				compatible = "fsl,imx93-mu", "fsl,imx8ulp-mu";
+				compatible = "fsl,imx93-mu-s4";
 				reg = <0x42440000 0x10000>;
 				interrupts = <GIC_SPI 24 IRQ_TYPE_LEVEL_HIGH>;
 				clocks = <&clk IMX93_CLK_MU2_B_GATE>;
@@ -685,6 +726,14 @@ wdog3: wdog@42490000 {
 				timeout-sec = <40>;
 			};
 
+			tpm3: pwm@424e0000 {
+				compatible = "fsl,imx7ulp-pwm";
+				reg = <0x424e0000 0x1000>;
+				clocks = <&clk IMX93_CLK_TPM3_GATE>;
+				#pwm-cells = <3>;
+				status = "disabled";
+			};
+
 			tpm4: pwm@424f0000 {
 				compatible = "fsl,imx7ulp-pwm";
 				reg = <0x424f0000 0x1000>;
@@ -696,6 +745,28 @@ tpm4: pwm@424f0000 {
 				status = "disabled";
 			};
 
+			tpm5: pwm@42500000 {
+				compatible = "fsl,imx7ulp-pwm";
+				reg = <0x42500000 0x1000>;
+				clocks = <&clk IMX93_CLK_TPM5_GATE>;
+				assigned-clocks = <&clk IMX93_CLK_TPM5>;
+				assigned-clock-parents = <&clk IMX93_CLK_24M>;
+				assigned-clock-rates = <24000000>;
+				#pwm-cells = <3>;
+				status = "disabled";
+			};
+
+			tpm6: pwm@42510000 {
+				compatible = "fsl,imx7ulp-pwm";
+				reg = <0x42510000 0x1000>;
+				clocks = <&clk IMX93_CLK_TPM6_GATE>;
+				assigned-clocks = <&clk IMX93_CLK_TPM6>;
+				assigned-clock-parents = <&clk IMX93_CLK_24M>;
+				assigned-clock-rates = <24000000>;
+				#pwm-cells = <3>;
+				status = "disabled";
+			};
+
 			i3c2: i3c-master@42520000 {
 				#address-cells = <3>;
 				#size-cells = <0>;
@@ -766,7 +841,7 @@ lpspi4: spi@42560000 {
 			};
 
 			lpuart3: serial@42570000 {
-				compatible = "fsl,imx93-lpuart", "fsl,imx8ulp-lpuart", "fsl,imx7ulp-lpuart";
+				compatible = "fsl,imx93-lpuart", "fsl,imx8ulp-lpuart";
 				reg = <0x42570000 0x1000>;
 				interrupts = <GIC_SPI 68 IRQ_TYPE_LEVEL_HIGH>;
 				clocks = <&clk IMX93_CLK_LPUART3_GATE>;
@@ -777,7 +852,7 @@ lpuart3: serial@42570000 {
 			};
 
 			lpuart4: serial@42580000 {
-				compatible = "fsl,imx93-lpuart", "fsl,imx8ulp-lpuart", "fsl,imx7ulp-lpuart";
+				compatible = "fsl,imx93-lpuart", "fsl,imx8ulp-lpuart";
 				reg = <0x42580000 0x1000>;
 				interrupts = <GIC_SPI 69 IRQ_TYPE_LEVEL_HIGH>;
 				clocks = <&clk IMX93_CLK_LPUART4_GATE>;
@@ -788,7 +863,7 @@ lpuart4: serial@42580000 {
 			};
 
 			lpuart5: serial@42590000 {
-				compatible = "fsl,imx93-lpuart", "fsl,imx8ulp-lpuart", "fsl,imx7ulp-lpuart";
+				compatible = "fsl,imx93-lpuart", "fsl,imx8ulp-lpuart";
 				reg = <0x42590000 0x1000>;
 				interrupts = <GIC_SPI 70 IRQ_TYPE_LEVEL_HIGH>;
 				clocks = <&clk IMX93_CLK_LPUART5_GATE>;
@@ -799,7 +874,7 @@ lpuart5: serial@42590000 {
 			};
 
 			lpuart6: serial@425a0000 {
-				compatible = "fsl,imx93-lpuart", "fsl,imx8ulp-lpuart", "fsl,imx7ulp-lpuart";
+				compatible = "fsl,imx93-lpuart", "fsl,imx8ulp-lpuart";
 				reg = <0x425a0000 0x1000>;
 				interrupts = <GIC_SPI 71 IRQ_TYPE_LEVEL_HIGH>;
 				clocks = <&clk IMX93_CLK_LPUART6_GATE>;
@@ -865,12 +940,6 @@ sai3: sai@42660000 {
 				status = "disabled";
 			};
 
-			mqs2: mqs2 {
-				compatible = "fsl,imx93-mqs";
-				gpr = <&wakeupmix_gpr>;
-				status = "disabled";
-			};
-
 			xcvr: xcvr@42680000 {
 				compatible = "fsl,imx93-xcvr";
 				reg = <0x42680000 0x800>,
@@ -894,7 +963,7 @@ xcvr: xcvr@42680000 {
 			};
 
 			lpuart7: serial@42690000 {
-				compatible = "fsl,imx93-lpuart", "fsl,imx8ulp-lpuart", "fsl,imx7ulp-lpuart";
+				compatible = "fsl,imx93-lpuart", "fsl,imx8ulp-lpuart";
 				reg = <0x42690000 0x1000>;
 				interrupts = <GIC_SPI 210 IRQ_TYPE_LEVEL_HIGH>;
 				clocks = <&clk IMX93_CLK_LPUART7_GATE>;
@@ -905,7 +974,7 @@ lpuart7: serial@42690000 {
 			};
 
 			lpuart8: serial@426a0000 {
-				compatible = "fsl,imx93-lpuart", "fsl,imx8ulp-lpuart", "fsl,imx7ulp-lpuart";
+				compatible = "fsl,imx93-lpuart", "fsl,imx8ulp-lpuart";
 				reg = <0x426a0000 0x1000>;
 				interrupts = <GIC_SPI 211 IRQ_TYPE_LEVEL_HIGH>;
 				clocks = <&clk IMX93_CLK_LPUART8_GATE>;
@@ -1161,7 +1238,7 @@ usdhc3: mmc@428b0000 {
 			};
 
 			epxp: epxp@4ae20000 {
-				compatible = "fsl,imx93-pxp-dma", "fsl,imx8ulp-pxp-dma";
+				compatible = "fsl,imx93-pxp-dma";
 				reg = <0x4ae20000 0x10000>;
 				interrupts = <GIC_SPI 173 IRQ_TYPE_LEVEL_HIGH>,
 					     <GIC_SPI 174 IRQ_TYPE_LEVEL_HIGH>;
@@ -1180,7 +1257,7 @@ cameradev: camera {
 				ranges;
 				status = "disabled";
 
-				isi_0: isi@4ae40000{
+				isi_0: isi@4ae40000 {
 					compatible = "fsl,imx93-isi", "fsl,imx8-isi";
 					reg = <0x4ae40000 0x10000>;
 					interrupts = <GIC_SPI 172 IRQ_TYPE_LEVEL_HIGH>;
@@ -1239,6 +1316,8 @@ parallel_csi: pcsi@4ac10070 {
 		};
 
 		gpio2: gpio@43810080 {
+			#address-cells = <1>;
+			#size-cells = <0>;
 			compatible = "fsl,imx93-gpio", "fsl,imx7ulp-gpio";
 			reg = <0x43810080 0x1000>, <0x43810040 0x40>;
 			gpio-controller;
@@ -1384,6 +1463,7 @@ ldb: ldb@4ac10020 {
 			#address-cells = <1>;
 			#size-cells = <0>;
 			compatible = "fsl,imx93-ldb";
+			reg = <0x4ac10020 0x4>;
 			clocks = <&clk IMX93_CLK_LVDS_GATE>;
 			clock-names = "ldb";
 			assigned-clocks = <&clk IMX93_CLK_MEDIA_LDB>;
@@ -1412,6 +1492,7 @@ ldb_ch0: endpoint {
 
 		ldb_phy: phy@4ac10024 {
 			compatible = "fsl,imx93-lvds-phy";
+			reg = <0x4ac10024 0x4>;
 			#address-cells = <1>;
 			#size-cells = <0>;
 			gpr = <&media_blk_ctrl>;
@@ -1469,8 +1550,6 @@ dsi_to_lcdif: endpoint {
 		};
 
 		lcdif: lcd-controller@4ae30000 {
-			#address-cells = <1>;
-			#size-cells = <0>;
 			compatible = "fsl,imx93-lcdif";
 			reg = <0x4ae30000 0x10000>;
 			interrupts = <GIC_SPI 176 IRQ_TYPE_LEVEL_HIGH>;
@@ -1490,10 +1569,9 @@ lcdif: lcd-controller@4ae30000 {
 			power-domains = <&media_blk_ctrl IMX93_MEDIABLK_PD_LCDIF>;
 			status = "disabled";
 
-			lcdif_disp: port@0 {
+			lcdif_disp: port {
 				#address-cells = <1>;
 				#size-cells = <0>;
-				reg = <0>;
 
 				lcdif_to_dsi: endpoint@0 {
 					reg = <0>;
@@ -1512,12 +1590,6 @@ lcdif_to_dpi: endpoint@2 {
 			};
 		};
 
-		usbphynop1: usbphynop1 {
-			compatible = "usb-nop-xceiv";
-			clocks = <&clk IMX93_CLK_USB_PHY_BURUNIN>;
-			clock-names = "main_clk";
-		};
-
 		usbotg1: usb@4c100000 {
 			compatible = "fsl,imx8mm-usb", "fsl,imx7d-usb";
 			reg = <0x4c100000 0x200>;
@@ -1539,12 +1611,6 @@ usbmisc1: usbmisc@4c100200 {
 			reg = <0x4c100200 0x200>;
 		};
 
-		usbphynop2: usbphynop2 {
-			compatible = "usb-nop-xceiv";
-			clocks = <&clk IMX93_CLK_USB_PHY_BURUNIN>;
-			clock-names = "main_clk";
-		};
-
 		usbotg2: usb@4c200000 {
 			compatible = "fsl,imx8mm-usb", "fsl,imx7d-usb";
 			reg = <0x4c200000 0x200>;
