Index: linux-tx93/arch/arm64/boot/dts/freescale/imx93.dtsi
===================================================================
--- linux-tx93.orig/arch/arm64/boot/dts/freescale/imx93.dtsi
+++ linux-tx93/arch/arm64/boot/dts/freescale/imx93.dtsi
@@ -810,6 +810,8 @@
 				clocks = <&clk IMX93_CLK_LPSPI3_GATE>,
 					 <&clk IMX93_CLK_BUS_WAKEUP>;
 				clock-names = "per", "ipg";
+				dmas = <&edma2 12 0 0>, <&edma2 13 0 1>;
+				dma-names = "tx","rx";
 				status = "disabled";
 			};
 
@@ -822,6 +824,8 @@
 				clocks = <&clk IMX93_CLK_LPSPI4_GATE>,
 					 <&clk IMX93_CLK_BUS_WAKEUP>;
 				clock-names = "per", "ipg";
+				dmas = <&edma2 14 0 0>, <&edma2 15 0 1>;
+				dma-names = "tx","rx";
 				status = "disabled";
 			};
 
@@ -1021,6 +1025,8 @@
 				clocks = <&clk IMX93_CLK_LPSPI5_GATE>,
 					 <&clk IMX93_CLK_BUS_WAKEUP>;
 				clock-names = "per", "ipg";
+				dmas = <&edma2 79 0 0>, <&edma2 80 0 1>;
+				dma-names = "tx","rx";
 				status = "disabled";
 			};
 
@@ -1033,6 +1039,8 @@
 				clocks = <&clk IMX93_CLK_LPSPI6_GATE>,
 					 <&clk IMX93_CLK_BUS_WAKEUP>;
 				clock-names = "per", "ipg";
+				dmas = <&edma2 81 0 0>, <&edma2 82 0 1>;
+				dma-names = "tx","rx";
 				status = "disabled";
 			};
 
@@ -1045,6 +1053,8 @@
 				clocks = <&clk IMX93_CLK_LPSPI7_GATE>,
 					 <&clk IMX93_CLK_BUS_WAKEUP>;
 				clock-names = "per", "ipg";
+				dmas = <&edma2 83 0 0>, <&edma2 84 0 1>;
+				dma-names = "tx","rx";
 				status = "disabled";
 			};
 
@@ -1057,6 +1067,8 @@
 				clocks = <&clk IMX93_CLK_LPSPI8_GATE>,
 					 <&clk IMX93_CLK_BUS_WAKEUP>;
 				clock-names = "per", "ipg";
+				dmas = <&edma2 85 0 0>, <&edma2 86 0 1>;
+				dma-names = "tx","rx";
 				status = "disabled";
 			};
 
