Index: linux-tx93/drivers/gpu/drm/imx/Kconfig
===================================================================
--- linux-tx93.orig/drivers/gpu/drm/imx/Kconfig
+++ linux-tx93/drivers/gpu/drm/imx/Kconfig
@@ -5,7 +5,7 @@ config DRM_IMX
 	select VIDEOMODE_HELPERS
 	select DRM_GEM_DMA_HELPER
 	depends on DRM && (ARCH_MXC || ARCH_MULTIPLATFORM || COMPILE_TEST)
-	depends on IMX_IPUV3_CORE || IMX_DPU_CORE || IMX8MM_LCDIF_CORE
+	depends on IMX_IPUV3_CORE || IMX_DPU_CORE || IMX8MM_LCDIF_CORE || IMX_LCDIFV3_CORE
 	help
 	  enable i.MX graphics support
 
