commit a31a661f9183340991f9955ad54b47f09141138d
Author: mb <mb@karo-electronics.de>
Date:   Wed Jun 19 08:20:25 2019 +0000

    add CAN controller support via SPI for MCP251x

diff --git a/arch/arm64/boot/dts/freescale/imx8mm-tx8m-1610-mipi-mb-can.dts b/arch/arm64/boot/dts/freescale/imx8mm-tx8m-1610-mipi-mb-can.dts
new file mode 100644
index 0000000..af97e97
--- /dev/null
+++ b/arch/arm64/boot/dts/freescale/imx8mm-tx8m-1610-mipi-mb-can.dts
@@ -0,0 +1,54 @@
+// SPDX-License-Identifier: (GPL-2.0 OR MIT)
+/*
+ * Copyright 2019 Lothar Waßmann <LW@KARO-electronics.de>
+ *
+ */
+
+#include <dt-bindings/interrupt-controller/irq.h>
+#include "imx8mm-tx8m-1610-mipi-mb.dts"
+
+/ {
+	mcp_clock: mcp-clock {
+		compatible = "fixed-clock";
+		#clock-cells = <0>;
+		clock-frequency = <16000000>;
+	};
+};
+
+&ecspi2 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_ecspi2>;
+	spi-max-frequency = <25000000>;
+	cs-gpios = <&gpio5 13 GPIO_ACTIVE_LOW>;
+	status = "okay";
+
+	can0: can@0 {
+		compatible = "microchip,mcp2515";
+		reg = <0>;
+		pinctrl-names = "default";
+		pinctrl-0 = <&pinctrl_can0>;
+		clocks = <&mcp_clock>;
+		interrupt-parent = <&gpio3>;
+		interrupts = <11 IRQ_TYPE_LEVEL_LOW>;
+		spi-max-frequency = <10000000>;
+	};
+};
+
+&iomuxc {
+	pinctrl_can0: can0grp {
+		fsl,pins = <
+			MX8MM_IOMUXC_NAND_CLE_GPIO3_IO5		0x156
+			MX8MM_IOMUXC_ECSPI2_SS0_GPIO5_IO13	0x156
+		>;
+
+	};
+
+	pinctrl_ecspi2: ecspi2grp {
+		fsl,pins = <
+			MX8MM_IOMUXC_ECSPI2_SCLK_ECSPI2_SCLK	0x156
+			MX8MM_IOMUXC_ECSPI2_MOSI_ECSPI2_MOSI	0x156
+			MX8MM_IOMUXC_ECSPI2_MISO_ECSPI2_MISO	0x156
+		>;
+
+	};
+};
