From e4d87c7bdd7e128a98dba65b1fbfa1585a39569e Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Lothar=20Wa=C3=9Fmann?= <LW@KARO-electronics.de>
Date: Tue, 3 Dec 2019 10:54:13 +0100
Subject: [PATCH 4/4] arm: dts: imx6ull: add imx6ull-txul-8013* files to
 Makefile

---
 arch/arm/boot/dts/Makefile | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/arch/arm/boot/dts/Makefile b/arch/arm/boot/dts/Makefile
index 291052a..033fafc 100644
--- a/arch/arm/boot/dts/Makefile
+++ b/arch/arm/boot/dts/Makefile
@@ -595,6 +595,9 @@ dtb-$(CONFIG_SOC_IMX6ULL) += \
 	imx6ull-9x9-evk-btwifi.dtb \
 	imx6ull-9x9-evk-btwifi-oob.dtb \
 	imx6ull-9x9-evk-ldo.dtb		\
+	imx6ull-txul-8013.dtb \
+	imx6ull-txul-8013-mainboard.dtb \
+	imx6ull-txul-8013-mb7.dtb \
 	imx6ulz-14x14-evk.dtb \
 	imx6ulz-14x14-evk-btwifi.dtb \
 	imx6ulz-14x14-evk-btwifi-oob.dtb \
-- 
2.7.4

