diff --git a/drivers/gpu/imx/lcdif/lcdif-common.c b/drivers/gpu/imx/lcdif/lcdif-common.c
index b3781270ad67..b20406b7fbf4 100644
--- a/drivers/gpu/imx/lcdif/lcdif-common.c
+++ b/drivers/gpu/imx/lcdif/lcdif-common.c
@@ -120,6 +120,7 @@ static int lcdif_enable_clocks(struct lcdif_soc *lcdif)
 			goto disable_disp_axi;
 	}
 
+pr_info("%s@%d: Enabling lcdif_pixel clk\n", __func__, __LINE__);
 	ret = clk_prepare_enable(lcdif->clk_pix);
 	if (ret)
 		goto disable_disp_apb;
@@ -138,6 +139,7 @@ static int lcdif_enable_clocks(struct lcdif_soc *lcdif)
 
 static void lcdif_disable_clocks(struct lcdif_soc *lcdif)
 {
+pr_info("%s@%d: Disabling lcdif_pixel clk\n", __func__, __LINE__);
 	clk_disable_unprepare(lcdif->clk_pix);
 
 	if (lcdif->clk_disp_axi)
@@ -449,8 +451,10 @@ void lcdif_set_mode(struct lcdif_soc *lcdif, struct videomode *vmode)
 	writel(CTRL1_FIFO_CLEAR, lcdif->base + LCDIF_CTRL1 + REG_CLR);
 
 	/* set pixel clock rate */
+pr_info("%s@%d: Disabling lcdif_pixel clk\n", __func__, __LINE__);
 	clk_disable_unprepare(lcdif->clk_pix);
 	clk_set_rate(lcdif->clk_pix, vmode->pixelclock);
+pr_info("%s@%d: Enabling lcdif_pixel clk\n", __func__, __LINE__);
 	clk_prepare_enable(lcdif->clk_pix);
 
 	/* config display timings */
